### DISCUSSION TITLE
Strange homing problem

### DISCUSSION THREAD

**Discussion by regpye** [arkypita/LaserGRBL#1377] [General]
I have a puzzling problem that I have not been able to work out, maybe someone here can advise me on what is going on and how to fix it?
I have built another diode laser engraver/cutter, it is my fifth one, so I am not a noobie and have some experience.
The machine uses Arduino UNO (genuine) and a Chinese clone CNC Shield v3. The firmware is the latest GRBL 1.1h. The shield has been setup with two Y drivers, cloning the A drive to be the other Y. The Z axis is also used
The software used is the latest LaserGRBL and I am using that on several other machines without any issues.
The machine has been tested and can move in all directions and the intended homing position is the front left hand corner and Z at max positive position.
I am able to travel positive and negative until I do a homing. The homing is travelling in the negative direction. After homing I can travel only in positive directions, regardless of which direction I choose.
I checked the wiring and was able to home successfully, but the X and Y will only travel in the positive direction, even if I select negative travel.
I disabled homing and soft limits, and tried again after resetting the computer and the machine.
After manually positioning the tool in the home position by hand and setting zero X and Y, I am able to move in both directions no problem. By setting the travel lengths to the numbers shown on the screen, I am able to travel and stop in the correct positions and using the home button, return to the zero X Y that I set.
Once I turn on homing the problem starts again, so it must be something to do with the homing setup I think. I checked the limit switches and they are all OK .
The homing direction is correct for my machine, but maybe it isn't?? I am homing to the front left hand corner. Very thing appears to be as it should, until homing has completed and then I can no longer move in any negative direction, except for the Z axis.

$0=10 (Step pulse time)
$1=25 (Step idle delay)
$2=0 (Step pulse invert)
$3=5 (Step direction invert)
$4=0 (Invert step enable pin)
$5=0 (Invert limit pins)
$6=0 (Invert probe pin)
$10=19 (Status report options)
$11=60.000 (Junction deviation)
$12=0.002 (Arc tolerance)
$13=0 (Report in inches)
$20=0 (Soft limits enable)  Can only enable if $21 =1
$21=0 (Hard limits enable)
$22=0 (Homing cycle enable) I turned this off to do further tests.
$23=3 (Homing direction invert)
$24=1200.000 (Homing locate feed rate)
$25=1000.000 (Homing search seek rate)
$26=144 (Homing switch debounce delay)
$27=3.000 (Homing switch pull-off distance)
$30=1000 (Maximum spindle speed)
$31=0 (Minimum spindle speed)
$32=0 (Laser-mode enable) I turned this off so I could test the laser without motion
$100=40.000 (X-axis travel resolution)
$101=40.000 (Y-axis travel resolution)
$102=80.000 (Z-axis travel resolution)
$110=10000.000 (X-axis maximum rate)
$111=10000.000 (Y-axis maximum rate)
$112=1000.000 (Z-axis maximum rate)
$120=1000.000 (X-axis acceleration)
$121=1000.000 (Y-axis acceleration)
$122=100.000 (Z-axis acceleration)
$130=840.000 (X-axis maximum travel)
$131=420.000 (Y-axis maximum travel)
$132=75.000 (Z-axis maximum travel)

Setting Value 	Mask 	        Invert X    Invert Y   Invert Z
0			00000000  	N	            N	           N
1			00000001	Y	            N	           N
2			00000010	N	            Y	           N
3			00000011	Y	            Y	           N
4			00000100	N	            N	           Y
5			00000101	Y	            N	           Y
6			00000110	N	            Y	           Y
7			00000111	Y	            Y	           Y

$3 – Direction port invert mask:binary
This setting inverts the direction signal for each axis. By default, Grbl assumes that the axes move in 
a positive direction when the direction pin signal is low, and a negative direction when the pin is 
high. Often, axes don't move this way with some machines. This setting will invert the direction pin 
signal for those axes that move the opposite way.
This invert mask setting works exactly like the step port invert mask and stores which axes to invert 
as bit flags. To configure this setting, you simply need to send the value for the axes you want to 
invert. Use the table above. For example, if want to invert the Y axis direction only, you'd send
$3=2 to Grbl and the setting should now read $3=2 (dir port invert mask:00000010)

**Comment by regpye** [comment]
I am not getting any replies, so I stripped down all the wiring and started again.
This time I only wired the stepper motors and the limit switches.
I turned off homing in the configuration to do the first test. The X and Y motors needed reversing, so instead of changing the configuration file, I physically changed the wiring so that the motors moved in the right directions. I made sure that $3=0
The test showed that the machine can move in all correct directions if I move the laser head to the "wanted" home position first by hand.
All working well so far.
Next I turned on the homing and set it as default of $23=0. The test showed that the machine was trying to home in the top right-hand corner, which is not where the limit switches are, so I stopped everything and changed $23=3 which should be the bottom left-hand corner.
I rebooted everything and did another test.  The homing went correctly, Z lifted up, double bounced and then the X an Y proceeded to the lower left-hand corner, The Y stopped first and waited until the X reached the limit, stopped and both did a double bounce. All seems normal, no warnings from LaserGRBL.
I then tried making a movement, and I could move both X and Y in the positive directions, but when I tried to go negative again, that is returning to home, the movements were both in the positive directions again.
This has got me really beat, I feel I have checked everything and just can't work out what is going on.

**Reply by regpye** [reply]
I may have nailed the problem down?
Because I was getting the repeated bad results from the hardware I was using, I decided to change all the electronics again.
I used a new Arduino UNO, programmed it with the original GRBL 1.1h and then I tried again using the same CNC shield.
Same problem. So next I also changed the CNC shield for a new one, checked all my wiring very carefully and tried again.
Same problem.
I was starting to think maybe the is something wrong in the software, so I opened up Universal G-code Sender and tried that.
Straight away everything worked, I could home and travel in all directions without any problems.
So next I closed UGS and opened up LaserGRBL again, and tried again. Back to the same problem, so to me it is something that LaserGRBL is doing.

**Reply by regpye** [reply]
Problem confirmed.  I loaded an old version of LaserGRBL and it works perfectly, so the new version has a jogging problem.

**Reply by arkypita** [reply]
> Problem confirmed. I loaded an old version of LaserGRBL and it works perfectly, so the new version has a jogging problem.

Buggy version?
Working version?

**Comment by arkypita** [comment]
I did not answer you before because unfortunately it is not possible to understand exactly the problem you describe.
People usually ask for help but don't explain the problem they have. On the contrary, you are explaining it too much, and I am confused.

You can make a simple schematic example:
1) I do this
2) the machine does this
3) instead it should do this

And maybe add some screenshots of the program and in particular of the jog commands that are shown in the left column? Or even a video?

**Reply by arkypita** [reply]
PS: i have homing in both my self-build and commercial engraver (ortur).
I use it and it work (in all direction).

**Reply by regpye** [reply]
Homing problem. After homing will only work in positive directions. 
OK before homing, can work in all directions using the arrow buttons..
If I run a file it works OK, just the jogging has the problem. using the arrow buttons.

A video on YouTube is here showing the problem.
Sorry it is not clear, but I did it in a hurry.
https://youtu.be/DTMGwkrw5e4

After loading an old version of LaserGRBL and testing, it works fine.
The configuration for the new version is shown below:

$0=10 (Step pulse time)
$1=25 (Step idle delay)
$2=0 (Step pulse invert)
$3=0 (Step direction invert)
$4=0 (Invert step enable pin)
$5=0 (Invert limit pins)
$6=0 (Invert probe pin)
$10=19 (Status report options)
$11=60.000 (Junction deviation)
$12=0.002 (Arc tolerance)
$13=0 (Report in inches)
$20=0 (Soft limits enable)
$21=0 (Hard limits enable)
$22=1 (Homing cycle enable)
$23=3 (Homing direction invert)
$24=1200.000 (Homing locate feed rate)
$25=1000.000 (Homing search seek rate)
$26=144 (Homing switch debounce delay)
$27=3.000 (Homing switch pull-off distance)
$30=1000 (Maximum spindle speed)
$31=0 (Minimum spindle speed)
$32=0 (Laser-mode enable)
$100=40.000 (X-axis travel resolution)
$101=40.000 (Y-axis travel resolution)
$102=80.000 (Z-axis travel resolution)
$110=10000.000 (X-axis maximum rate)
$111=10000.000 (Y-axis maximum rate)
$112=1000.000 (Z-axis maximum rate)
$120=1000.000 (X-axis acceleration)
$121=1000.000 (Y-axis acceleration)
$122=100.000 (Z-axis acceleration)
$130=840.000 (X-axis maximum travel)
$131=420.000 (Y-axis maximum travel)
$132=75.000 (Z-axis maximum travel)

I used the same configuration for the old version of LaserGRBL.

**Reply by arkypita** [reply]
The video shows the problem, but it does not give me enough information to understand what the cause is.

What you are showing is not repeatable on any of my machines and, considering the user of LaserGRBL (60'000 active users) I am more inclined to think that it is a problem with your machine.

> **And maybe add some screenshots of the program and in particular of the jog commands that are shown in the left column?**

As you can see in my screenshot the jog command sent from LaserGRBL is the same before and after homing.
What about yours?

![image](https://user-images.githubusercontent.com/8782035/112278612-51729400-8c83-11eb-9596-abb23ffa87e8.png)

**Reply by regpye** [reply]
I have done that test and noticed that the blue cross that represents the tool position moves in the same way as the machine does.  I will try and make a video of it happening so it will be clearer for you.

![test homing](https://user-images.githubusercontent.com/2367937/112296553-f5d8f280-8ce4-11eb-93fc-3892f366e3e8.jpg)

**Reply by regpye** [reply]
![test-homing](https://user-images.githubusercontent.com/2367937/112297140-713aa400-8ce5-11eb-876d-fbded3757aaf.jpg)

**Reply by arkypita** [reply]
> that the blue cross that represents the tool position moves in the same way as the machine does

The blue cross moves based on what the machine says.

> I opened up Universal G-code Sender and tried that.
> Straight away everything worked, I could home and travel in all directions without any problems.

I have no idea what codes UGS sends for jogging, as far as I know it can use simple G0 commands: the evidence proves nothing.

LaserGRBL uses the $J command as defined by the [grbl v1.1 documentation](https://github.com/gnea/grbl/wiki/Grbl-v1.1-Jogging).

![image](https://user-images.githubusercontent.com/8782035/112299795-70c7ec00-8c98-11eb-99d4-28c7afecf697.png)

**The image you uploaded shows that LaserGRBL sends the correct commands:**
- X840.0 to jog towards the limit of the X axis
- X0.0 to jog in direction of zero point
- same for Y

**Comment by arkypita** [comment]
I believe that your machine put itself in negative space after homing.
If the machine is in negative space going in direction of zero or going in direction of +840 move in the same direction.

**Reply by arkypita** [reply]
Look at the numbers in top-right of lasergrbl UI after homing (if you stopped putting cropped images!)

**Reply by regpye** [reply]
OK, I will make a screen capture video tomorrow (it is late here now) and that will show all the screen in full motion.

I think I understand what you are saying, but how could this problem be overcome? 

I have already re-programmed a new UNO board, and changed the CNC shield board, checked the config settings with another version of LaserGRBL (the old version doesn't give this problem)

What would you suggest I try next? Maybe after seeing the next video will give some clues?

**Reply by regpye** [reply]
OK, I couldn't sleep so I went out to the workshop and made a video.  I hope it shows what we need to know about this problem. Thanks for your patience.

https://youtu.be/HHMerlSqbVs

**Reply by arkypita** [reply]
After homing, at 1:20 https://youtu.be/HHMerlSqbVs?t=80 the engraver goes in negative space.

**Reply by arkypita** [reply]
You should edit grbl [config.h](https://github.com/gnea/grbl/blob/bfb67f0c7963fe3ce4aaf8a97f9009ea5a8db36e/grbl/config.h#L126) before compiling and uncomment this line:

// #define HOMING_FORCE_SET_ORIGIN // Uncomment to enable.

**Reply by arkypita** [reply]
> I have already re-programmed a new UNO board, and changed the CNC shield board, checked the config settings with another version of LaserGRBL (the old version doesn't give this problem)

I don't understand why you keep claiming it is something that depends on LaserGRBL: LaserGRBL does things correctly.

**Reply by regpye** [reply]
I have tried to change the config.h file and made the change that you mentioned (removed the comments marks //
and now have #define HOMING_FORCE_SET_ORIGIN
When I tried to compile I got some errors.
I tried using the machine again with no changes at all, just used LaserGRBL 2.8.5 a very old version, and it works perfectly, but doesn't have the new features that I need for the laser..
That version doesn't show the coordinated on the top right hand screen like the newer versions, so making a video of it wont show the problem.

I will try again to see if I can compile GRBL successfully.

**Reply by arkypita** [reply]
> I tried using the machine again with no changes at all, just used LaserGRBL 2.8.5 a very old version

I believe that this version does not have "continuous jog" so it move step by step (using different jog command).
You can disable continuous jog in the newest LaserGRBL version if you want, but best is fix your firmware.

PS. you can flash grbl directly from LaserGRBL (tools menu) v1.1h custom XY homing

**Reply by regpye** [reply]
**Problem partly solved.**
I was able to reset the coordinates using this command:

G10 L2 P1 X0Y0Z0

Then I noticed that the screen had changed and the jogging control had not been turned on.  As soon as I turned on continuous jogging, problem came back again.
I will have to try and compile GRBL again, not sure why I am getting errors.

**Reply by regpye** [reply]
**Success!**
I followed your advise and re-compiled GRBL, worked out the compiling problem and was successful in updating the firmware.
Tested it out and all is working as it should do.
Thanks so much Arkypita for all your help and a wonderful program too.
